### PR TITLE
Feature/collection new

### PR DESCRIPTION
**Pull request by ekate**
Updated collection version. Bind nyucores and collections, using Hydra is_part_of predicate. 
It's a temporary solution. Hydra7 uses default fedora map of predicates, if I want to add a predicate I have to overwrite the map. I am not sure how overwrite will work in our production environment so used the default one
If I understand it correctly in Hydra 9-10 we will be able use DC properties directly, so will probably redo after upgrade
Also the PR includes new collection - Indian Ocean Postcards. The collection has to be private so I add restriction to solr parameters. The restriction is based on collection parameter "discoverable"

### CODE REVIEW

**Comment by jgpawletko** [comment]
Hi @ekate ,
I haven't finished reviewing your PR, but I have some initial questions and suggestions:

**Is the logic correct here?**
https://github.com/NYULibraries/ichabod/blob/83891a755908f7f77427b9253a773829d3c8c1ab/app/controllers/catalog_controller.rb#L200

```
      if !can?(:edit, collection)
```
Based on you comments, should it be the following instead?
```
      if can?(:edit, collection)
```
Do you have a test that checks if collection discovery restrictions are working?

**please consider this change**
https://github.com/NYULibraries/ichabod/pull/230/files#diff-b00d6b9614b5cfd3c4fa3875da6c1978R19
from:
```
    if (is_restricted?)
      raise CanCan::AccessDenied.new('You are not authorized to view this item')
    else
      respond_with(@item)
    end
```
to:
```
   raise CanCan::AccessDenied.new('You are not authorized to view this item') if is_restricted?
   respond_with(@item)
```

**please consider this change**
https://github.com/NYULibraries/ichabod/pull/230/files#diff-b00d6b9614b5cfd3c4fa3875da6c1978R120
from:
```
  def is_restricted?
   collection=Collection.find(@item.collection.pid)  unless(@item.collection.nil?)
   if (!@item.collection.nil?)
     return true if (@item.collection.discoverable=='0'&&!can?(:edit, collection))
   end
   false
 end
```
to:
```
  def is_restricted?
   unless @item.collection.nil?
     collection = Collection.find(@item.collection.pid)
     return true if ((@item.collection.discoverable == ‘0’) && (!can?(:edit, collection)))
   end
   false
  end
```

**concerns about `discoverable`**
We're using a string to determine whether something is discoverable.
This feels brittle to me.  For example, you're using numeric string-literals, 
e.g. `'0'`, whereas I see that the [`NYUCore` factory](https://github.com/NYULibraries/ichabod/blob/fb1d0cc2a1e50c3a35a4daecde06aa54816046ff/spec/factories/nyucores.rb#L35
) uses an alphabetic 
string literal, e.g., `'Y'`.

I think we need to agree on what we're using to determine discoverability.
I suggest using boolean values.   If we absolutely have to use strings, then let's
agree on a controlled set of strings.  

Also, could you please implement a `Collection#discoverable?` predicate method?
The method hides implementation details and allows us to change this code:
```
return true if (@item.collection.discoverable == '0'&& !can?(:edit, collection))
```
into this code:
```
return true if (@item.collection.discoverable? && !can?(:edit, collection))
```

**please consider this change**
https://github.com/NYULibraries/ichabod/pull/230/files#diff-c83c2cebbb04f7c73d37edf794b7c936R64
from:
```
  def destroy
    authorize! :destroy, params[:id]
    @collection = Collection.find(params[:id])
    if @collection.nyucores.size>0
      respond_to do |format|
        format.html { redirect_to collections_url, notice: 'Collection has assosiated records and can not be deleted' }
      end
    else
      @collection.destroy
      respond_to do |format|
        format.html { redirect_to collections_url, notice: 'Collection was successfully deleted.' }
      end
    end
  end
```
to:
```
# can tighten up logic a bit, and require that collection size is exactly zero
  def destroy
    authorize! :destroy, params[:id]
    @collection = Collection.find(params[:id])
    if @collection.nyucores.size == 0
      @collection.destroy
  	 msg = 'Collection was successfully deleted.'
    else	
  	 msg = 'Collection has associated records and can not be deleted'
    end
    flash[:notice] = msg
    redirect_to collections_url
  end
```

Please note: the proposed code changes **have not been tested**. I might have some typos.
Thanks!

**Comment by jgpawletko** [comment]
Hi @ekate,
Thank you for answering some of my previous comments (in person).
My only other comment is about certain constructs that feel brittle to me.

For example,  [this line](https://github.com/NYULibraries/ichabod/pull/230/files#diff-956e46b33fb5307990c4e1a4b5fd86ccR201) from the method `show_only_discoverable_records`:
```
solr_params[:fq] << '-is_part_of_ssim:'+"info\\:fedora/#{collection.pid.gsub(":","\\:")}"
```

I like how you encapsulated the logic in `show_only_discoverable_records`, but
the line above is doing several things:
1.) it's adding collections to exclude from the search results 
2.) it's building a fedora identifier from the collection pid
3.) it's escaping the fedora identifier for submission to Solr

Please let me know if I've misread your code.
It would be good, in general, to break these functions out into separate methods
so that we're not repeating ourselves, and we could reuse the functions throughout the
code.  If we ever needed to change the way the escaping works, we would only
need to change one function, not search and replace through the code.

**Comment by chrpr** [comment]
Hi @ekate 

First off, this is _awesome_ work. I really like how you've set this up. A few comments / questions below.

* I'm not quite clear what you mean by "Hydra7 uses default fedora map of predicates, if I want to add a predicate I have to overwrite the map." Can you link to the code you're talking about? You should be able to define and use predicates easily, in the same manner we're doing in the Nyucore model itself. And you've already done so for :discoverable, right?

*  If I understand correctly, discoverable and collection are now _only_ part of the collection model? If so, do we want to pull them into a different part of the YAML file to make that clear? Currently, the are processed by the nyucore model as if they are also valid there. But maybe we want to be able to inherit them to the item level later, in which case we should leave them?  Maybe we iterate, and at a future point we actually isolate our definition of all the collection fields in CollectionsDatastream instead of re-using NyucoreDatastream.

* We probably need some documentation here -- maybe an update to the readme (or an added "Collections" readme), or even just an update to our [Starting up](https://github.com/NYULibraries/ichabod/wiki/Starting-Up) wiki page? I figured it out, but it took a while...
  * Start by wiping any existing data in both test and dev database. Easiest to do so by running `bundle exec rake jetty:clean`
  * Create necessarily collections (eg 'Research Guides') at localhost:3000/collections before reloading data
  * Reload data, eg: `bundle exec rake ichabod:load['lib_guides','./ingest/libguides.xml']`
  * While we're in there, should we also add the correct reload syntax for each collection?

**Comment by ekate** [comment]
@chrpr If I understand it correctly in Hydra 7 you can only use Fedora specific predicates to define relationship between object. Basically if I want to use ActiveFedora methods when working with relationships I need to use predicates designated for it in Fedora mapping. I can change it but it means changing the fedora config file.

**Comment by chrpr** [comment]
@ekate Thanks for the clarification. I didn't realize you were talking about relationships between fedora classes. Yes, you're absolutely correct about that. Good to know that we can use other properties to manage inter-fedora relationships in Hydra 9/10. 

As I said, this looks great! I don't think there's much else that needs addressing here. I still think we may later decide to create CollectionsDatastream for the collection fields, but we can iterate there. 

And if you'd like, I'd be happy to help with the documentation suggestions I made.

Thanks!

**Comment by jgpawletko** [comment]
Hi @ekate, 
Is this PR ready to go?

Hi @NYULibraries/hydra,
Are there any remaining issues/concerns with this PR?

Thanks!
Joe

**Comment by eshadatta** [comment]
@jgpawletko ,

I don't have any comments. Just responding to your query.

File: app/models/collection.rb [hunk 1]
```diff
@@ -13,8 +16,8 @@ class Collection < ActiveFedora::Base
   REQUIRED_FIELDS = [ :title, :discoverable ]
   ID_PREFIX="ichabodcollection"
 
-
   validates :title, presence: true
+  #validates :title, uniqueness: { solr_name: "desc_metadata__title_tesim" }
```
Comment (da70) [inline app/models/collection.rb]: Recommend removing this if no longer valid.

File: app/models/nyucore.rb [hunk 2]
```diff
@@ -106,11 +108,7 @@ def to_solr(solr_doc = {})
       # https://github.com/projecthydra/active_fedora/blob/889aa962a326ad9e8302ada3237193221ad2feb5/lib/active_fedora/indexing.rb
       solr_doc = ds.to_solr(solr_doc)
     end
-    Solrizer.insert_field(solr_doc, "collection", collections, :facetable, :displayable)
+    #Solrizer.insert_field(solr_doc, "collection", collections, :facetable, :displayable)
```
Comment (da70) [inline app/models/nyucore.rb]: Recommend removing this if no longer valid.

File: app/controllers/catalog_controller.rb [hunk 3]
```diff
@@ -2,19 +2,21 @@
...
 class CatalogController < ApplicationController
-
   include Blacklight::Catalog
   include Hydra::Controller::ControllerBehavior
-  # These before_filters apply the hydra access controls
-  # before_filter :enforce_show_permissions, :only=>:show
-  # This applies appropriate access controls to all solr queries
-  # CatalogController.solr_search_params_logic += [:add_access_controls_to_solr_params]
+  
+  #add check for descoverability on collection level
```
Comment (da70) [inline app/controllers/catalog_controller.rb]: "descoverability" -> "discoverability"

File: app/controllers/catalog_controller.rb [hunk 4]
```diff
@@ -2,19 +2,21 @@
...
+  #(see https://github.com/projectblacklight/blacklight/wiki/Extending-or-Modifying-Blacklight-Search-Behavior)
+  #as we are currently using older version of Blacklight, additional method couldn't be placed to search_builder class
+  #So tempoirary added it to the controller itself
```
Comment (da70) [inline app/controllers/catalog_controller.rb]: "tempoirary" -> "temporarily"

File: app/controllers/collections_controller.rb [hunk 5]
```diff
@@ -61,9 +61,15 @@ def update
   def destroy
     authorize! :destroy, params[:id]
     @collection = Collection.find(params[:id])
-    @collection.destroy
-    respond_to do |format|
-      format.html { redirect_to collections_url, notice: 'Collection was successfully deleted.' }
+    if @collection.nyucores.size>0
+      respond_to do |format|
+        format.html { redirect_to collections_url, notice: 'Collection has assosiated records and can not be deleted' }
```
Comment (da70) [inline app/controllers/collections_controller.rb]: "assosiated" -> "associated"

File: spec/controllers/collections_controller_spec.rb [hunk 6]
```diff
@@ -91,10 +91,19 @@
   end
 
   describe "DELETE destroy" do
-    let!(:collection) { create(:collection, { :title => "LION" }) }
-    it 'should delete an existing collection record and return to the main page' do
+    context 'when collection has no assosiated record' do
```
Comment (da70) [inline spec/controllers/collections_controller_spec.rb]: "assosiated" -> "associated"

File: spec/controllers/collections_controller_spec.rb [hunk 7]
```diff
@@ -91,10 +91,19 @@
...
+      let!(:collection) { create(:collection, { :title => "LION" }) }
+      it 'should delete an existing collection record and return to the main page' do
           expect { delete :destroy, id: collection }.to change(Collection, :count).by(-1)
           expect response.should redirect_to collections_url
+      end
+    end
+    context 'when collection has  assosiated record' do
```
Comment (da70) [inline spec/controllers/collections_controller_spec.rb]: "assosiated" -> "associated"

File: app/helpers/catalog_helper.rb [hunk 8]
```diff
@@ -10,12 +10,20 @@ def render_external_links(args)
     field_name = args[:field]
     begin
       url_presenter = UrlPresenter.new(document, field_name)
-      links = url_presenter.urls.collect do |url|
-        link_to(url.text, url.value, {target: '_blank'})
-      end
-      links.join(field_value_separator).html_safe
+        links = url_presenter.urls.collect do |url|
```
Comment (da70) [inline app/helpers/catalog_helper.rb]: Whitespace/indentation in this block seems off.

File: app/helpers/catalog_helper.rb [hunk 9]
```diff
@@ -10,12 +10,20 @@ def render_external_links(args)
...
+          link_to(url.text, url.value, {target: '_blank'})
```
Comment (da70) [inline app/helpers/catalog_helper.rb]: Whitespace/indentation in this block seems off.

File: spec/ichabod/resource_set/base_spec.rb [hunk 10]
```diff
@@ -230,6 +271,10 @@ module ResourceSet
               expect(nyucore.edit_groups).to eq editors.map(&:to_s).unshift(*original_editors.map(&:to_s))
             end
           end
+          it 'should should assign edit groups to the parent collection' do
```
Comment (da70) [inline spec/ichabod/resource_set/base_spec.rb]: There is a repeated "should" here.

**Comment by da70** [comment]
@jgpawletko and everyone, I did a very quick, partial, and shallow scan of the code after reading everyone's comments and just pointed out a few very minor typos and style issues in Github comments.  None of which need to be done except for "assosiated" -> "associated" since it's in error messages and also tests output, and hence potentially consumable later by automated processes.  Since Kate's on vacation I could go ahead and fix them for her.  Don't know if the protocol/workflow allows for that.

I agree with Joe that the `discoverable` magic values should be made consistent.  We've run into time-consuming problems before due to inconsistent use of string vs. numeric representation.  The `discoverable?` method is also a nice idea.

**Comment by ekate** [comment]
@chrpr from this tutorial I understood that you can be more flexibale when define predicates in Fedora 4 
https://github.com/projecthydra/hydra/wiki/Lesson---Define-Relationships-Between-Objects
However we need to look ate how the new RDF Fedora module works

**Comment by ekate** [comment]
@chrpr @jgpawletko I was thinking what to do with discoverable filed as when I attempt to make it boolean in the YAML  it doesn't work. For now I just used 'Y/N'. I was thinking probably @chrpr  is correct and we should have a separate YAML for collections. Collections are our internal objects and we will not re-use them as we do with NYU core, correct ? But we still need discoverable field for items I think

**Comment by ekate** [comment]
@jgpawletko @chrpr @eshadatta @da70 Sorry for the delay in response. I had internet connection/charger issues. I've implemented changes Joe recommended (but not the boolean field) but just realized I need to add tests for new methods so the coverage not decreased. I will submit the final changes (including the ones David recommended) tomorrow.
Thank you for your comments !!

**Comment by jgpawletko** [comment]
Hi @NYULibraries/hydra ,

I'm ready to merge this PR.

Please comment on this thread if you have any concerns, especially re: rebasing branches.

**If I don't hear from anyone by 5:00 p.m., I will merge and, if Jenkins and Travis are happy,
will push to Production,**

Thank you!
Best-
Joe

**Pull request merged**